cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java

```diff
@@ -101,65 +122,50 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
-                                          @Nullable Map<String, String> arguments,
-                                          @Nullable ClassLoader classLoader,
-                                          DatasetClassLoaderProvider classLoaderProvider,
-                                          @Nullable Iterable<? extends Id> owners,
-                                          AccessType accessType)
+  public <T extends Dataset> T getDataset(final Id.DatasetInstance datasetInstanceId,
+                                          @Nullable final Map<String, String> arguments,
+                                          @Nullable final ClassLoader classLoader,
+                                          final DatasetClassLoaderProvider classLoaderProvider,
+                                          @Nullable final Iterable<? extends Id> owners,
+                                          final AccessType accessType)
     throws DatasetManagementException, IOException {
-    T dataset = super.getDataset(datasetInstanceId, arguments, classLoader, classLoaderProvider, owners, accessType);
-    writeLineage(datasetInstanceId, dataset, accessType);
-    writeUsage(datasetInstanceId, owners);
-    return dataset;
-  }
-
-  private void writeUsage(Id.DatasetInstance datasetInstanceId, @Nullable Iterable<? extends Id> owners) {
-    if (null == owners) {
-      return;
-    }
+    Principal principal = authenticationContext.getPrincipal();
     try {
-      runtimeUsageRegistry.registerAll(owners, datasetInstanceId);
+      // For system, skip authorization and lineage (user program shouldn't allow to access system dataset CDAP-6649)
+      // For non-system dataset, always perform authorization and lineage.
+      AuthorizationEnforcer enforcer;
+      DefaultDatasetRuntimeContext.DatasetAccessRecorder accessRecorder;
+      if (Id.Namespace.SYSTEM.equals(datasetInstanceId.getNamespace())) {
+        enforcer = SYSTEM_NAMESPACE_ENFORCER;
+        accessRecorder = SYSTEM_NAMESPACE_ACCESS_RECORDER;
+      } else {
+        enforcer = authorizationEnforcer;
+        accessRecorder = new BasicDatasetAccessRecorder(datasetInstanceId, accessType, owners);
+      }
+
+      return DefaultDatasetRuntimeContext.execute(enforcer, accessRecorder, principal,
+                                                  datasetInstanceId.toEntityId(), new Callable<T>() {
+          @Override
+          public T call() throws Exception {
+            return LineageWriterDatasetFramework.super.getDataset(datasetInstanceId, arguments, classLoader,
+                                                                  classLoaderProvider, owners, accessType);
+          }
+        });
+    } catch (IOException | DatasetManagementException e) {
+      throw e;
     } catch (Exception e) {
-      LOG.warn("Failed to register usage of {} -> {}", owners, datasetInstanceId, e);
+      throw new DatasetManagementException("Failed to craete dataset instance: " + datasetInstanceId, e);
```

Comment:
The usage registration won't throw exception in the try block above. Usage registration is now done during dataset constructor/method call, and the exception is handling inside those call (failure in usage recording won't cause failure of the operation).